File: javascript/matrix/matrix.js
function extractRows(arr) {
  return arr;
}

function extractColumns(arr) {
  return arr[0].map((e, i) => arr.map(r => r[i]));
}

function extractArray(str) {
  const rows = str.split('\n');
  const strArr = rows.map(r => r.split(' '));
  const intArr = strArr.map(r => r.map(e => parseInt(e, 10)));
  return intArr;
}

function Matrix(str) {
  this.str = str;
  const arr = extractArray(str);
  this.rows = extractRows(arr);
  this.columns = extractColumns(arr);
}

module.exports = Matrix;

File: ruby/matrix/matrix.rb
# frozen_string_literal: true

# Store a 2D matrix
class Matrix
  attr_reader :rows

  def initialize(string)
    @rows = parse(string)
  end

  def columns
    @rows.transpose
  end

  private

  def parse(string)
    string.split(/\n+/).map(&method(:parse_line))
  end

  def parse_line(line)
    line.split.map(&:to_i)
  end
end
